lms/services/blackboard_api/_schemas.py

```diff
@@ -8,21 +8,37 @@
 from lms.validation._base import RequestsResponseSchema
 
 
+class _FileSchema(Schema):
+    """Shared base schema for Blackboard file dicts."""
+
+    class Meta:
+        unknown = EXCLUDE
+
+    id = fields.Str(required=True)
+    display_name = fields.Str(data_key="name", required=True)
+    updated_at = fields.Str(data_key="modified", required=True)
+
+    @post_load
+    def post_load(self, data, **_kwargs):  # pylint:disable=no-self-use
+        data["id"] = f"blackboard://content-resource/{data['id']}/"
+        return data
```

Comment:
The path I ended up following in the end was to stick the the existing design as on master, which is that the schema does all the work of deserializing the data. So `_FileSchema` handles excluding unknown fields (like `downloadUrl`), renaming the `name` and `modified` fields, and even wrapping the `id` in a `blackboard://...` URL. And then `BlackboardListFilesSchema` also handles popping the `"results"` key. The actual service method is then just a one-liner

Reply:
Is it no possible to *not* overwrite ['id'] and add this as a separate field (or keep the original id as a different field, `lms_id` or similar)

If I was using the service I'd expect to use the ID coming from one API method as the ID needed for another API call

Reply:
The problem is that this is what happens:

1. Our list files proxy API returns a list of file dicts to our frontend: `[{"id": "blackboard://...", ...}, { ... }, { ... }]`
2. The user selects a file to use for the assignment
3. Our frontend submits the file's `blackboard://...` ID to our backend as the `document_url` field and it gets saved in the DB

The fact that the `document_url` submitted by the frontend starts with `blackboard://...` rather than `https://...` is the only way that the backend knows that it's a Blackboard file.

So we kind of do need to overwrite the ID.

We could do the overwrite in the service method rather than in the schema.

Or we could do the overwrite in the `list_files()` proxy API _view_ rather than in the service or schema.

But without changing the frontend the ID does need to be overwritten before we send it to the frontend.

Reply:
If this is massaging the data for the JS it might belong to the view but in the other hand you'll end up spilling the `blackboard://...` knowledge over more places, that's probably a worse outcome.

It's keeping the orignal ID On `lms_id` or similar feasible or marshmallow would make that difficult?

Reply:
I actually think it's best for the schema to do all the work. The reason is that it means the schema is the only thing that knows anything about the Blackboard files API's response format, and the schema's tests are the only tests that need to know anything about it. So there's a [fake Blackboard API response in `_schemas_test.py`](https://github.com/hypothesis/lms/blob/9b021fa5832d906548438518abab998881047ce2/tests/unit/lms/services/blackboard_api/_schemas_test.py#L91-L110) and that's the only place in the tests where we need a fake Blackboard API response. The service and view just return what the schema returned directly, and their tests just need to assert that they returned whatever the schema returned.

If we split the responsibility of validating and mutating the Blackboard API responses between the schema, service and view, then knowledge about the API response format has to be duplicated in multiple places in the code and tests. The worse case would be if the schema just returned the Blackboard API response unmodified, and then the service made some changes to it (renaming and omitting fields), and then the view made some more changes (changing the ID's into URLs). In that case the schemas tests would need a fake Blackboard API response, the service's tests would also need a fake Blackboard API response, and the view's tests would need a fake Blackboard API response as it would look after being partially mutated by the service. And the tests would be like this for a reason: because the responsibility for validating and mutating those responses and the knowledge about the response format has actually been split between those three places. If the work is only split between two places (schema and service, or schema and view) you get a lesser version of the same problem.

That's why I think it's actually best for the schema to do "all the work" here: when you try it the code and tests actually work out simpler.

I don't think that will always be the case when you have a schema, service and view. But in this one particular case it happens to work out that way for now.

* * *

We could move the original ID to an `"lms_id"` item (`post_load()` is just a Python function, it can do whatever we want) but it wouldn't change much: the schema just returns these dicts to the service which directly returns them to the view which directly returns them to the frontend, and the frontend isn't expecting an `"lms_id"` field, it would just ignore it (and not send it back to the backend). So nothing would be able to use the `"lms_id"` field.

We could also _add_ a separate `"document_url"` field instead of changing the `"id"` field but again the frontend would ignore a `"document_url"` and just return the `"id"` (in a form field named `document_url`!) so it would break as the backend wouldn't be able to tell that it's a Blackboard file.

I think we might want a frontend change to allow us to include `document_url`'s in the file dicts instead of changing the `id` into a URL, because changing the `id` is just a bit weird. But I'll have to coordinate with the frontend team and see what's possible

Reply:
Started a Slack thread about potential frontend changes: https://hypothes-is.slack.com/archives/C1M8NH76X/p1623836733025000